### PR TITLE
Loop unrolling

### PR DESCRIPTION
**Pull request by Jacarte**
Loop unrolling mutator for empty-returning loops

cc @fitzgen

### CODE REVIEW

File: crates/wasm-mutate/src/mutators/codemotion/mutators/loop_unrolling.rs [hunk 1]
```diff
@@ -0,0 +1,246 @@
+//! This mutator selects a random `if` construction in a function and swap its branches.
+//! Since this mutator preserves the original semantic of the input Wasm,
+//! before the mutated if structure is encoded, a "negation" of the previous operand
+//! in the stack is written. The "negation" is encoded with a `i32.eqz` operator.
```
Comment (fitzgen) [inline crates/wasm-mutate/src/mutators/codemotion/mutators/loop_unrolling.rs:4]: Looks like this was copy-pasted from the if/else swapping mutator.

File: crates/wasm-mutate/src/mutators/codemotion/mutators/loop_unrolling.rs [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+                                def += 1;
+                            }
+
+                            to_fix.insert(idx, Instruction::BrTable(jmpfix.into(), def));
+                        }
+                        _ => {}
+                    }
+                }
+
+                // println!("To fix {:?}", to_fix);
```
Comment (fitzgen) [inline crates/wasm-mutate/src/mutators/codemotion/mutators/loop_unrolling.rs:141]:
```suggestion
```

File: crates/wasm-mutate/src/mutators/codemotion/mutators/loop_unrolling.rs [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        &self,
+        ast: &Ast,
+        nodeidx: usize,
+        body: &[usize],
+        newfunc: &mut Function,
+        operators: &Vec<OperatorAndByteOffset>,
+        input_wasm: &'a [u8],
+        ty: &wasmparser::TypeOrFuncType,
+    ) -> crate::Result<()> {
+        if self.loop_to_mutate == nodeidx && !self.mutated.get() {
```
Comment (fitzgen) [inline crates/wasm-mutate/src/mutators/codemotion/mutators/loop_unrolling.rs:185]: Why do we need `self.mutated`? Is there a reason we can't rely on the node index alone?

File: crates/wasm-mutate/src/mutators/codemotion/mutators/loop_unrolling.rs [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+        _: &crate::ModuleInfo,
+        rnd: &mut rand::prelude::SmallRng,
+        ast: &Ast,
+        locals: &[(u32, ValType)],
+        operators: &Vec<OperatorAndByteOffset>,
+        input_wasm: &'a [u8],
+    ) -> crate::Result<Function> {
+        // Select the if index
+        let mut newfunc = Function::new(locals.to_vec());
+        let empty_retuurning_loops = self.get_empty_returning_loops(ast);
```
Comment (fitzgen) [inline crates/wasm-mutate/src/mutators/codemotion/mutators/loop_unrolling.rs:235]:
```suggestion
        let empty_returning_loops = self.get_empty_returning_loops(ast);
```

**Comment by Jacarte** [comment]
Comments addressed :)

File: crates/wasm-mutate/src/mutators/codemotion/mutators/loop_unrolling.rs [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+        let mut to_fix = HashMap::new();
+
+        match &nodes[nodeidx] {
+            Node::Loop { body: _, ty, range } => {
+                let chunk = &operators[range.start + 1 /* skip the loop instruction */..range.end];
+                newfunc.instruction(&Instruction::Block(map_block_type(*ty)?));
+                for (idx, (op, _)) in chunk.iter().enumerate() {
+                    match op {
+                        Operator::Block { .. } => {
+                            stack.push(State::Block);
```
[the remark this answers is not in the document]
Comment (Jacarte) [reply crates/wasm-mutate/src/mutators/codemotion/mutators/loop_unrolling.rs]: Done ;)

**Review by fitzgen — APPROVED** [review]
Thanks!

**Pull request merged**